### ISSUE TITLE
container:quickstart does't use webappPostProcess

### ISSUE THREAD

**Issue by mrArkwright** [earldouglas/xsbt-web-plugin#373]
We're using `webappPostProcess` to add some css sources that were generated from less files to the webapp. Those files are missing when the container ist started with `quickstart`.

This behaviour seems unavoidable the way `quickstart` currently works. However it would be nice to have an option to make `quickstart` work on a copy of the web sources in the target folder (similar to `container:start` but without the jars). This way it would become possible to invoke `webappPostProcess`. Of course this would mean that we loose the ability to see changes in the web sources without restarting the container. I personally could live with that. If it's only an option and not the default behaviour this also would't break any existing projects. Alternatively a second command (`container:quickstart2` or so) could be introduced that incorporates this behaviour.

I'm aware that `container:start` doesn't have this issue. However we're using xsbt-web-plugin in a project with a lot of scala sources and packaging takes about 60 seconds. That makes it infeasible to use `start` in place of `quickstart`.

I'm also happy to help and open a PR if we can agree on a solution. I already experimented with this a little bit.

**Comment by earldouglas** [comment]
Rather than using `webappPostProcess` for this, could you use `resourceGenerators` to produce your `.css` files prior to compilation, and place the output in `./src/main/webapp/`?

```scala
resourceGenerators in Compile +=
  Def.task({
    import sys.process._
    val styleLess = (sourceDirectory in Compile).value / "less" / "style.less"
    val styleCss  = (sourceDirectory in Compile).value / "webapp" / "style.css"
    s"./node_modules/.bin/lessc ${styleLess} ${styleCss}".!
    Seq.empty[File] //(styleCss)
  }).taskValue
```

To use this with triggered execution (e.g. `~container:quickstart`), you would want to cache the output (see the `cacheify` function in `WebAppPlugin` for an example).  Otherwise, the generated `.css` will be detected as a source change (since `src/main/webapp` is a source directory) and the container would restart over and over again.

**Comment by mrArkwright** [comment]
I thought about something like that as well. The problem would be that `./src/main/webapp/` would now contain a build artifact. style.css would have to go in the .gitignore. It's also hard to anticipate which files might be added to that in the future. I'm also not sure right now if mixing sources and artifacts in a folder would cause other problems as well.

**Comment by earldouglas** [comment]
All great points.  Using `resourceGenerators` like this would be at best a hacky workaround.

> However it would be nice to have an option to make quickstart work on a copy of the web sources in the target folder

This option seems like it could be a reasonable solution.  If we change `quickstart` to use the output of `webappPrepare` for the container's webapp path, then we would get the desired behavior.

Currently, `quickstart` does two things:

1. Passes the full runtime classpath of the project (including the project's compiled classes and upstream dependencies) to the container
2. Skips the `webappPrepare` step, and instead points the container to `src/main/webapp` instead of `target/webapp`.

We could just get rid of the second half of this.  It would slow down `quickstart` only as much as it takes to copy `src/main/webapp` to `target/webapp`.  Could this work for your build?  We would still skip the packaging step, avoiding the heavy lift of making and copying `.jar` files around for the project and all its dependencies.

**Comment by earldouglas** [comment]
On second thought, this wouldn't be enough.  `webappPrepare` causes project code to be packaged and dependencies to be copied to `target/webapp/lib`.

We would have to either split up `webappPrepare` into two separate tasks, or add a new `webappPreProcess` task.

**Comment by earldouglas** [comment]
What about this?

1. Add `webappPrePrepare` that simply copies `src/main/webapp` to `target/webapp`
1. Add `wepappPreProcess` that a user can configure (just like `webappPostProcess`), and depends on `webappPrePrepare`
1. Update `webappPrepare` to depend on `wepappPreProcess`
1. Update `quickstart` to depend on `wepappPreProcess`, and use `target/webapp` instead of `src/main/webapp`, just like `start` does today

**Comment by mrArkwright** [comment]
Yes, this is more or less what I'm doing right now. One problem though: If `quickstart` works on `target/webapp` jetty picks up any present jars that might have been copied there by an earlier `start`. My solution is to have a separate `webappPrepareQuick` command that copies the relevant files to `target/webapp-quick` and then invokes `webappPostProcess`.

Do you think we should just alter the way way quickstart works? We could also introduce a settings key to switch between the current and the new behaviour.

**Comment by earldouglas** [comment]
Great point.  We need separate target directories for pre-prepare and prepare tasks.  Here's a first attempt at this:

https://github.com/earldouglas/xsbt-web-plugin/compare/webapp-pre-process?expand=1

Feel free to try this out by updating your xsbt-web-plugin version to `ISSUE-373`, and let me know if it works for you:

*project/plugins.sbt:*

```scala
addSbtPlugin("com.earldouglas" % "xsbt-web-plugin" % "ISSUE-373")
```

Here's how I'm testing this:

*build.sbt:*

```scala
webappPreProcess := {
  webappDir: File =>
    import sys.process._
    val styleLess = (sourceDirectory in Compile).value / "less" / "style.less"
    val styleCss  = webappDir / "style.css"
    s"./node_modules/.bin/lessc ${styleLess} ${styleCss}".!
}
```

**Comment by mrArkwright** [comment]
Works like a charm!

However in this scenario where `webappPePrepare` and `webappPrepare` work on different directories it doesn't make sense that `webappPrepare` depends on `webappPePrepare` right? Both tasks could just copy their files from `sourceDirectory`. Or am I missing something?

Also couldn't we just invoke `webappPostProcess` at the end of `webappPePrepare`? That way we wouldn't need the second function `webappPreProcess`.

**Comment by earldouglas** [comment]
> However in this scenario where webappPePrepare and webappPrepare work on different directories it doesn't make sense that webappPrepare depends on webappPePrepare right? Both tasks could just copy their files from sourceDirectory.

Almost:

1. pre-prepare copies `src/main/webapp`, and then runs pre-process
2. prepare takes the output of the above from `target/webapp-pre`, copies it to `target/webapp`, and then does the rest of the webapp prep (copying `.jars`, running post-process)

> Also couldn't we just invoke webappPostProcess at the end of webappPrePrepare? That way we wouldn't need the second function webappPreProcess.

We could, though that would change the semantics of post-process.  Currently, post-process is assumed to run after the rest of webapp prep is finished, which includes copying over the compiled project artifacts and all the dependent `.jar` files.

**Comment by earldouglas** [comment]
Based on your suggestion, here's an alternative implementation that might work better for us:

1. Have `quickstart` copy `src/main/webapp` to `target/webapp-quick`
2. Run `webappPostProcess` on `target/webapp-quick`
3. Don't change `webappPrepare` from how it works today

I like this approach because it doesn't break any existing tests, and doesn't introduce any new tasks.

**Comment by mrArkwright** [comment]
Yes, I like this approach better as well. It also avoids a two level caching situation (copying from sources to `webapp-pre` and then to `webapp`) which I find hard to reason about (but might not impose an actual problem).

**Comment by earldouglas** [comment]
Done.  Give this one a try with version `ISSUE-373-2`:

*project/plugins.sbt:*

```scala
addSbtPlugin("com.earldouglas" % "xsbt-web-plugin" % "ISSUE-373-2")
```

*build.sbt:*

```scala
webappPostProcess := {
  webappDir: File =>
    import sys.process._
    val styleLess = (sourceDirectory in Compile).value / "less" / "style.less"
    val styleCss  = webappDir / "style.css"
    s"./node_modules/.bin/lessc ${styleLess} ${styleCss}".!
}
```

**Comment by mrArkwright** [comment]
Perfect! Thanks a lot! I appreciate it!

A minor detail: Since `webappPrePrepare` isn't a prerequisite for `webappPrepare` anymore but only a reduced version of `webappPrepare` specially tailored to `quickstart` maybe a better name would be `webappPrepareQuick` or something along the lines?

**Comment by earldouglas** [comment]
Yep: https://github.com/earldouglas/xsbt-web-plugin/pull/374/files#diff-4673190a00d0ae3102a35c0c8d02325fR30

**Comment by mrArkwright** [comment]
Nice!

`webappPostProcess` seems to get invoked twice in `webappPrepareTask` though.

**Comment by earldouglas** [comment]
Oops, nice catch.  Should now be fixed in version `ISSUE-373-3`.

**Comment by mrArkwright** [comment]
No further objections from my side. I'm more than happy with this solution (:

**Comment by earldouglas** [comment]
Great.  I'll add a test and release it as 4.2.0.  Thanks for your help!

**Comment by earldouglas** [comment]
This is now released in version 4.2.0.  Enjoy!
